k2/csrc/log.cuh

```diff
@@ -123,13 +127,32 @@ class Logger {
 
 class Voidifier {
  public:
-  __host__ __device__ void operator&(const Logger &)const {}
+  K2_CUDA_HOSTDEV void operator&(const Logger &) const {}
 };
 
 }  // namespace internal
 
 }  // namespace k2
 
+/**
+ * @brief A static assertion
+ *
+ * @param[in] COND the compile-time boolean expression that must be true
+ * @param[in] MSG  an error message if exp is false
+ *
+ * @note `static_assert` is supported by both of host and device.
+ *
+ * @code{.cpp}
+ * K2_STATIC_ASSERT(DEFINED_SHAPE % DEFINED_X == 0);
+ * @endcode
+ */
+#define K2_STATIC_ASSERT4(COND, MSG) \
+  typedef char static_assertion_##MSG[(!!(COND))*2-1]
```

Comment:
I guess in c++11 and later, people would use `static_asstert`. This kind of trick is commonly seen in pre-c++11 code.

Reply:
The reason is the old static_assert requires a msg. It's a problem for current code. The new static_assert is ready since c++17.
I am fine to just use static_assert. @danpovey please make a choice.

```cpp
static_assert ( bool_constexpr , message )		(since C++11)
static_assert ( bool_constexpr )		(since C++17)
```

Reply:
Mm.   Perhaps just define K2_STATIC_ASSERT(x) static_assert(x, "")

Reply:
maybe we can add a macro like ` K2_STATIC_ASSERT` represents `static_assert(bool, "")`?